hal/src/stm32f2xx/i2c_hal.c

```diff
@@ -1,726 +1,740 @@
+    NVIC_InitStructure.NVIC_IRQChannelPreemptionPriority = 0;
+    NVIC_InitStructure.NVIC_IRQChannelSubPriority = 0;
+    NVIC_InitStructure.NVIC_IRQChannelCmd = ENABLE;
+    NVIC_Init(&NVIC_InitStructure);
+
+    if(mode != I2C_MODE_MASTER)
+    {
+        NVIC_InitStructure.NVIC_IRQChannel = i2cMap[i2c]->I2C_EV_IRQn;
+        NVIC_Init(&NVIC_InitStructure);
+    }
+
+    i2cMap[i2c]->I2C_InitStructure.I2C_Mode = I2C_Mode_I2C;
+    i2cMap[i2c]->I2C_InitStructure.I2C_DutyCycle = I2C_DutyCycle_2;
+    i2cMap[i2c]->I2C_InitStructure.I2C_OwnAddress1 = address << 1;
+    i2cMap[i2c]->I2C_InitStructure.I2C_Ack = I2C_Ack_Enable;
+    i2cMap[i2c]->I2C_InitStructure.I2C_AcknowledgedAddress = I2C_AcknowledgedAddress_7bit;
+    i2cMap[i2c]->I2C_InitStructure.I2C_ClockSpeed = i2cMap[i2c]->I2C_ClockSpeed;
+
+    /*
+        From STM32 peripheral library notes:
+        ====================================
+        If an error occurs (ie. I2C error flags are set besides to the monitored
+        flags), the I2C_CheckEvent() function may return SUCCESS despite
+        the communication hold or corrupted real state.
+        In this case, it is advised to use error interrupts to monitor
+        the error events and handle them in the interrupt IRQ handler.
+     */
+    I2C_ITConfig(i2cMap[i2c]->I2C_Peripheral, I2C_IT_ERR, ENABLE);
+
+    if(mode != I2C_MODE_MASTER)
+    {
+        I2C_ITConfig(i2cMap[i2c]->I2C_Peripheral, I2C_IT_EVT | I2C_IT_BUF, ENABLE);
+    }
+
+    /* Enable the I2C peripheral */
+    I2C_Cmd(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+
+    /* Apply I2C configuration after enabling it */
+    I2C_Init(i2cMap[i2c]->I2C_Peripheral, &i2cMap[i2c]->I2C_InitStructure);
+
+    i2cMap[i2c]->I2C_Enabled = true;
+}
+
+void HAL_I2C_End(HAL_I2C_Interface i2c, void* reserved)
+{
+    if(i2cMap[i2c]->I2C_Enabled != false)
+    {
+        I2C_Cmd(i2cMap[i2c]->I2C_Peripheral, DISABLE);
+
+        i2cMap[i2c]->I2C_Enabled = false;
+    }
+}
+
+uint32_t HAL_I2C_Request_Data(HAL_I2C_Interface i2c, uint8_t address, uint8_t quantity, uint8_t stop, void* reserved)
+{
+    uint32_t _millis;
+    uint8_t bytesRead = 0;
+
+    // clamp to buffer length
+    if(quantity > BUFFER_LENGTH)
+    {
+        quantity = BUFFER_LENGTH;
+    }
+
+    /* Send START condition */
+    I2C_GenerateSTART(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+
+    _millis = HAL_Timer_Get_Milli_Seconds();
+    while(!I2C_CheckEvent(i2cMap[i2c]->I2C_Peripheral, I2C_EVENT_MASTER_MODE_SELECT))
+    {
+        if(EVENT_TIMEOUT < (HAL_Timer_Get_Milli_Seconds() - _millis))
+        {
+            /* SW Reset the I2C Peripheral */
+            HAL_I2C_SoftwareReset(i2c);
+
+            return 0;
+        }
+    }
+
+    /* perform blocking read into buffer */
+    uint8_t *pBuffer = i2cMap[i2c]->rxBuffer;
+    uint8_t numByteToRead = quantity;
+
+	if (quantity == 1)
+		/* Disable Acknowledgement */
+		I2C_AcknowledgeConfig(i2cMap[i2c]->I2C_Peripheral, DISABLE);

```

Comment:
Disabling ACK if quantity == 1 here (no comment as to why that is needed before Slave Address send), but disabling ACK if numByteToRead (aka quantity) == 2 below later on receive.  Can we explain this difference better in comments?